astropy/time/core.py

```diff
@@ -495,6 +495,8 @@ def _get_time_fmt(self, val, val2, format, scale,
         else:
             formats = [(format, self.FORMATS[format])]
 
+        assert formats
```

Comment:
@taldcroft - since we were just discussing a similar thing, this is the type of "assert that my assumptions hold" I meant in https://github.com/astropy/astropy/pull/9340#discussion_r333988907 - what if due to some bug `formats` is an empty list? Should definitely not happen, but who knows. The subsequent code would mean that what gets returned is `None`, which is not really what one wants!

Aside: not quite sure it is needed here, but it does show the intent.

Reply:
The distinction I use is, if the precondition is something the user should enforce, I raise a ValueError (or whatever) with helpful messages. If the precondition is obviously true as the result of immediately-visible astropy code, I don't check at all. But if the precondition is, as here, something astropy *ought* to enforce, but as a result of interactions of code and data structures that could easily be changed without thought to this function, then I use assert. Especially if the failure mode is hard to understand: failure to accept the input, signaled by an error raised by the attempt to index format[0] while constructing a different exception with a confusing message.

Essentially: ValueErrors (or TypeErrors or whatever) for errors caused by users or external circumstances; assertions for errors caused by people modifying astropy itself. Plus, especially given this rule, "assert" documents that the code will implode if ever there are no permitted formats left when one gets here.